src/main/kotlin/com/exactpro/th2/processor/core/message/protobuf/controller/CradleMessageGroupState.kt

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2022-2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.processor.core.message.protobuf.controller
+
+import com.exactpro.th2.common.grpc.AnyMessage
+import com.exactpro.th2.common.grpc.MessageGroup
+import com.exactpro.th2.common.grpc.MessageGroupBatch
+import com.exactpro.th2.common.message.logId
+import com.exactpro.th2.common.message.toJson
+import com.exactpro.th2.common.utils.message.book
+import com.exactpro.th2.common.utils.message.sessionGroup
+import com.exactpro.th2.common.utils.message.timestamp
+import com.exactpro.th2.processor.core.message.AbstractMessageState
+import com.exactpro.th2.processor.core.message.AbstractMessageState.Companion.StateKey
+import com.exactpro.th2.processor.utility.check
+import com.exactpro.th2.processor.utility.compareTo
+import com.google.protobuf.Timestamp
+import com.google.protobuf.util.Timestamps
+
+internal class CradleMessageGroupState(
+    private val startTime: Timestamp,
+    private val endTime: Timestamp,
+    private val kinds: Set<AnyMessage.KindCase>,
+    bookToGroups: Map<String, Set<String>>
+): AbstractMessageState<MessageGroupBatch, MessageGroup>(
+    kinds.size,
+    bookToGroups
+) {
+    override fun MessageGroup.toStateKey(batch: MessageGroupBatch): StateKey {
+        check()
+
+        val firstMessage = requireNotNull(messagesList.first()) {
```

Comment:
The `first` method will throw an exception itself if the `messageList` is empty. Probably, you want to use `firstOrNull` method here instead